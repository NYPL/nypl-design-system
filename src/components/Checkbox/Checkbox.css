.checkbox {
  align-items: baseline;
  display: flex;

  &__input {
    margin-right: var(--space-xs);
  }
}
