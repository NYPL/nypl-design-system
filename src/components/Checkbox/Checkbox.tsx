import * as React from "react";
import bem from "../../utils/bem";
import HelperErrorText from "../HelperErrorText/HelperErrorText";
import Label from "../Label/Label";
import generateUUID from "../../helpers/generateUUID";

export interface CheckboxProps {
  /** The checkbox's label.  This will serve as the text content for a `<label>` element if `showlabel` is true, or an "aria-label" if `showLabel` is false. */
  labelText: string;
  /** Additional attributes to pass to the `<input>` tag */
  attributes?: { [key: string]: any };
  /** When using the Checkbox as a "controlled" form element, you can specify the Checkbox's checked state using this prop. Learn more about controlled and uncontrolled form fields: https://goshakkk.name/controlled-vs-uncontrolled-inputs-react/ */
  checked?: boolean;
  /** className you can add in addition to 'input' */
  className?: string;
  /** Adds the 'disabled' prop to the input when true */
  disabled?: boolean;
  /** Helper for modifiers array; adds 'errored' styling */
  errored?: boolean;
  /** Optional string to populate the HelperErrorText for error state */
  errorText?: string;
  /** Optional string to populate the HelperErrorText for standard state */
  helperText?: string;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** The name prop indicates into which group of checkboxes this checkbox belongs.  If none is specified, 'default' will be used */
  name?: string;
  /** The action to perform on the `<input>`'s onChange function  */
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  /** Offers the ability to show the checkbox's label onscreen or hide it. Refer to the `labelText` property for more information. */
  showLabel: boolean;
  /** Populates the value of the input */
  value?: string;
}

export const onChangeDefault = () => {
  return;
};

const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  (props, ref?) => {
    const {
      checked,
      className,
      disabled,
      errored,
      errorText,
      helperText,
      id,
      labelText,
      name,
      showLabel,
      value,
    } = props;

    const attributes = props.attributes || {};
    const modifiers = props.modifiers ? props.modifiers : [];
    const checkboxID = id || generateUUID();
    const onChange = props.onChange || onChangeDefault;

    const footnote = errored ? errorText : helperText;

    if (!showLabel) {
      attributes["aria-label"] =
        labelText && footnote ? `${labelText} - ${footnote}` : labelText;
    } else {
      if (footnote) attributes["aria-describedby"] = `${checkboxID}-helperText`;
    }

    if (errored) {
      modifiers.push("error");
    }

    return (
      <>
        <input
          {...(checked !== undefined
            ? {
                checked: checked,
                onChange: onChange,
              }
            : {
                defaultChecked: false,
              })}
          className={bem("checkbox", modifiers, "input", [className])}
          disabled={disabled}
          id={checkboxID}
          name={name || "default"}
          ref={ref}
          type="checkbox"
          value={value}
          {...attributes}
        />
        {labelText && showLabel && (
          <Label htmlFor={checkboxID}>{labelText}</Label>
        )}
        {footnote && showLabel && (
          <HelperErrorText isError={errored} id={id + `-helperText`}>
            {footnote}
          </HelperErrorText>
        )}
      </>
    );
  }
);

Checkbox.displayName = "Checkbox";

export default Checkbox;
