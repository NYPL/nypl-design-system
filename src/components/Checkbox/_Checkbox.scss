$checkbox-size: to-rem(22px);
$checkmark-color: $focus;
$checkmark-color-error: $error;
$checkmark-color-disabled: $gray-dark;

.input {
  &__checkbox {
    -webkit-appearance: none;
    border: 2px solid $black;
    border-radius: 3px;
    display: inline-block;
    height: $checkbox-size;
    outline: none;
    position: relative;
    vertical-align: middle;
    width: $checkbox-size;

    &:before {
      content: "";
      display: block;
      height: $checkbox-size;
      left: 50%;
      position: absolute;
      top: 50%;
      transform: translate(-50%, -50%);
      width: $checkbox-size;
    }

    &:checked {
      border: 2px solid $focus;

      &:before {
        background: url('data:image/svg+xml;utf8,<svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path fill="#{url-friendly-colour($checkmark-color)}" d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"/></svg>')
          50% 50% no-repeat;
      }
    }

    + label {
      display: inline-block;
      font-size: var(--font-size-0);
      font-weight: 300;
      margin-bottom: 0px;
      margin-left: var(--space-xs);
      vertical-align: middle;

      + .helper-text {
        margin-bottom: 0;
        margin-left: calc(22px + var(--space-xs));
        margin-top: var(--space-xs);
      }
    }

    &--error {
      border-color: $error;

      + label {
        color: $error;

        + .helper-text {
          color: $error;
        }
      }

      &:checked {
        border-color: $error;

        &:before {
          background: url('data:image/svg+xml;utf8,<svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path fill="#{url-friendly-colour($checkmark-color-error)}" d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"/></svg>')
            50% 50% no-repeat;
        }
      }
    }

    &:disabled {
      border-color: $gray-medium;

      &:checked {
        border-color: $gray-medium;

        &:before {
          background: url('data:image/svg+xml;utf8,<svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path fill="#{url-friendly-colour($checkmark-color-disabled)}" d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"/></svg>')
            50% 50% no-repeat;
        }
      }

      + label {
        color: $gray-dark;
        font-style: italic;

        + .helper-text {
          color: $gray-dark;
          font-style: italic;
        }
      }
    }
  }
}
