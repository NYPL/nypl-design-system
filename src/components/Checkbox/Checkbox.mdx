import { Meta, Canvas, Description, ArgTypes } from "@storybook/blocks";

import * as CheckboxStories from "./Checkbox.stories";

<Meta of={CheckboxStories} />

# Checkbox

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.1.0`    |
| Latest            | `1.5.0`    |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [Checked](#checked)
- [Indeterminate](#indeterminate)
- [Focused](#focused)
- [Errored](#errored)
- [Disabled](#disabled)
- [Helper Text](#helper-text)
- [Error Text](#error-text)
- [With JSX Element Label](#with-jsx-element-label)

## Overview

<Description of={CheckboxStories} />

The `Checkbox` component renders a checkbox input element. The `Checkbox` component should be used to allow users to select one or more options from a list of choices.

Please note that the examples on this page show the `Checkbox` component in isolation. It is recommended to use the `Checkbox` component inside the
DS `CheckboxGroup` component. The `CheckboxGroup` component will handle all the states and data management.

## Component Props

<Canvas of={CheckboxStories.WithControls} />

<ArgTypes of={CheckboxStories.WithControls} />

## Accessibility

The `Checkbox` component renders `<label>` and `<input type="checkbox" />`
elements that are associated with each other; the label has a `for` attribute
with the value of the `input`'s `id` element. If there is helper or invalid text,
the `<input>` element will have a `aria-describedby` attribute with the value of
the helper or invalid text.

When `showLabel` is set to false, the `labelText` value will be set to the
`<input>`'s `aria-label` attribute.

Resources:

- [W3C WAI ARIA Checkbox Example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/checkbox/checkbox-1/checkbox-1.html)
- [a11ymatters Accessible Checkbox](https://www.a11ymatters.com/pattern/checkbox/)
- [MDN ARIA: checkbox role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/checkbox_role)
- [Chakra UI Checkbox](https://chakra-ui.com/docs/components/form/checkbox)

## Checked

Note that the `isChecked` property in this example is set to `true` and clicking
on it won't remove the checkmark; the `isChecked` prop must now be controlled
and removed manually by the parent component that controls this state.

<Canvas of={CheckboxStories.Checked} />

## Indeterminate

Note: This state only changes the icon through the `isIndeterminate` prop. This
does not affect whether the checkbox is checked. The `isIndeterminate` prop must
now be controlled and removed by the parent component in order to remove this state.

<Canvas of={CheckboxStories.Indeterminate} />

## Additional States

<Canvas of={CheckboxStories.AdditionalStates} />

## With JSX Element Label

This is useful when you want to add dynamic content to the label or add
a layout to the label. View the `CheckboxGroup` documentation for this
usage.

<Canvas of={CheckboxStories.WithJSXElementLabel} />
