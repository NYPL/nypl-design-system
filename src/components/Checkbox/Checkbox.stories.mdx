import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Preview,
  Description,
} from "@storybook/addon-docs/blocks";
import { withDesign } from "storybook-addon-designs";
import { getCategory } from "../../utils/componentCategories";

import Checkbox from "./Checkbox";
import { onChangeDefault } from "./Checkbox";

<Meta
  title={getCategory("Checkbox")}
  component={Checkbox}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url:
        "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=11895%3A658",
    },
  }}
  argTypes={{
    onChange: { table: { disable: true } },
    checked: { table: { disable: true } },
  }}
/>

# Checkbox

<Description of={Checkbox} />

This component renders a Checkbox button form element. See below for configuration information. Note that the `id` property, while optional, will
cause an accessibility violation if none is present.

<Preview withToolbar>
  <Story
    name="Checkbox"
    args={{
      labelText: "Test Label",
      showLabel: true,
      id: "test_id",
    }}
  >
    {args => <Checkbox {...args} />}
  </Story>
</Preview>

<ArgsTable story="Checkbox" />

## Checked

<Canvas>
  <Checkbox
    checked
    showLabel={true}
    onChange={onChangeDefault}
    labelText="I am checked"
  ></Checkbox>
</Canvas>

## Focused

<Canvas>
  <Checkbox
    name="testFocus"
    labelText="Click or tab to the Radio to see its focus state"
    showLabel={true}
  ></Checkbox>
</Canvas>

## Errored

<Canvas>
  <Checkbox
    errored
    showLabel={true}
    labelText="I am in an error state"
  ></Checkbox>
</Canvas>

## Disabled

<Canvas>
  <Checkbox disabled showLabel={true} labelText="I am disabled"></Checkbox>
</Canvas>

## Helper Text

<Canvas>
  <Checkbox
    name="testHelperText"
    showLabel={true}
    labelText="I have helper text"
    helperText="I am the helper text for this Radio"
  ></Checkbox>
</Canvas>

## Error Text

<Canvas>
  <Checkbox
    name="testErrorText"
    showLabel={true}
    errored
    labelText="I have error text"
    errorText="I am the error text for this Radio"
  ></Checkbox>
</Canvas>
