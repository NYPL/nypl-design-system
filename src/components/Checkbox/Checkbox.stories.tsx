import { Box, HStack, VStack } from "@chakra-ui/react";
import Heading from "../Heading/Heading";
import type { Meta, StoryObj } from "@storybook/react";

import Checkbox from "./Checkbox";

const argsBooleanType = (defaultValue = false) => ({
  control: { type: "boolean" },
  table: { defaultValue: { summary: defaultValue } },
});

const meta: Meta<typeof Checkbox> = {
  title: "Components/Form Elements/Checkbox",
  component: Checkbox,
  argTypes: {
    className: { control: false },
    helperText: { control: { type: "text" } },
    id: { control: false },
    invalidText: { control: { type: "text" } },
    isChecked: argsBooleanType(),
    isDisabled: argsBooleanType(),
    isIndeterminate: argsBooleanType(),
    isInvalid: argsBooleanType(),
    isRequired: argsBooleanType(),
    labelText: { control: { type: "text" } },
    name: { control: { type: "text" } },
    onChange: { control: false },
    showHelperInvalidText: argsBooleanType(true),
    showLabel: argsBooleanType(true),
    value: { control: { type: "text" } },
  },
};

export default meta;
type Story = StoryObj<typeof Checkbox>;

/**
 * Main Story for the Checkbox component. This must contains the `args`
 * and `parameters` properties in this object.
 */
export const WithControls: Story = {
  args: {
    className: undefined,
    helperText: "This is the helper text!",
    id: "checkbox_id",
    invalidText: "This is the error text :(",
    isChecked: undefined,
    isDisabled: false,
    isIndeterminate: false,
    isInvalid: false,
    isRequired: false,
    labelText: "Test Label",
    name: "test_name",
    onChange: undefined,
    showHelperInvalidText: true,
    showLabel: true,
    value: "1",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=11895%3A658",
    },
    jest: ["Checkbox.test.tsx"],
  },
};

// The following are additional Checkbox example Stories.
export const Checked: Story = {
  render: () => (
    <Checkbox id="isChecked" labelText="I am checked" isChecked value="1" />
  ),
};

export const Indeterminate: Story = {
  render: () => (
    <Checkbox
      id="indeterminate"
      labelText="I am indeterminate"
      isIndeterminate
    />
  ),
};

export const AdditionalStates: Story = {
  render: () => (
    <VStack align="left" spacing="l">
      <Box>
        <Heading level="h3" size="heading6">
          Focused
        </Heading>
        <Checkbox
          id="focused"
          labelText="Click or tab to the Checkbox to see its focus state"
        />
      </Box>
      <Box>
        <Heading level="h3" size="heading6">
          Errored
        </Heading>
        <HStack spacing="l">
          <Checkbox id="invalid" isInvalid labelText="I am in an error state" />
          <Checkbox
            id="invalid-checked"
            isInvalid
            isChecked
            labelText="I am checked in an error state"
          />
        </HStack>
      </Box>
      <Box>
        <Heading level="h3" size="heading6">
          Disabled
        </Heading>
        <HStack spacing="l">
          <Checkbox id="disabled" isDisabled labelText="I am disabled" />
          <Checkbox
            id="disabled-checked"
            isDisabled
            isChecked
            labelText="I am checked and disabled"
          />
        </HStack>
      </Box>
      <Box>
        <Heading level="h3" size="heading6">
          With Helper Text
        </Heading>
        <Checkbox
          id="helpertext"
          name="testHelperText"
          labelText="I have helper text"
          helperText="I am the helper text for this Checkbox"
        />
      </Box>
      <Box>
        <Heading level="h3" size="heading6">
          With Errored Text
        </Heading>
        <Checkbox
          id="invalid-text"
          invalidText="I am the error text for this Checkbox"
          isInvalid
          name="testinvalidText"
          labelText="I have error text"
        />
      </Box>
    </VStack>
  ),
};

export const LabelUsingJSXElements: Story = {
  render: () => (
    <Checkbox
      id="jsx-label"
      labelText={
        <span>
          Arts and <strong>Entertainment</strong>
        </span>
      }
      name="jsxElementLabel"
    />
  ),
};
