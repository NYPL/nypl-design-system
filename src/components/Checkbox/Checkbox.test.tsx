import { expect } from "chai";
import * as Enzyme from "enzyme";
import * as React from "react";
import { stub, spy } from "sinon";
import generateUUID from "../../helpers/generateUUID";

import Checkbox from "./Checkbox";

describe("Checkbox", () => {
  let container;
  let changeHandler;
  let clickHander;
  let generateUUIDSpy;

  before(() => {
    clickHander = stub();
    changeHandler = stub();
    generateUUIDSpy = spy(generateUUID);
    container = Enzyme.mount(
      <Checkbox
        id="inputID"
        attributes={{ onClick: clickHander }}
        onChange={changeHandler}
        labelText="Test Label"
        showLabel={false}
      ></Checkbox>
    );
  });

  it("Renders with appropriate 'aria-label' attribute and value when 'showLabel' prop is set to false", () => {
    const labelText = container.prop("labelText");
    expect(container.find("input").prop("aria-label")).to.equal(labelText);
  });

  it("Renders without crashing", () => {
    expect(container.find("input").exists()).to.equal(true);
  });

  it("The checkbox element is an input with type='checkbox'", () => {
    expect(container.find("input").prop("type")).to.equal("checkbox");
  });

  it("The checkbox element's ID is set properly using the value passed to it.", () => {
    expect(container.find("input").prop("id")).to.equal("inputID");
  });

  it("Allows user to pass in additional attributes", () => {
    container.simulate("click");
    expect(clickHander.callCount).to.equal(1);
  });

  it("Changing the value calls the onChange handler", () => {
    container = Enzyme.mount(
      <Checkbox
        id="onChangeTest"
        onChange={changeHandler}
        labelText="onChangeTest Lab"
        showLabel={true}
        checked
      ></Checkbox>
    );
    container.find("input").simulate("change", { target: { value: "Hello" } });
    expect(changeHandler.callCount).to.equal(1);
  });

  it("Renders with appropriate 'aria-label' attribute and value when 'showLabel' prop is set to false and 'helperText' has been passed", () => {
    container = Enzyme.mount(
      <Checkbox
        id="onChangeTest"
        onChange={changeHandler}
        labelText="onChangeTest Lab"
        showLabel={false}
        helperText="The helper text."
        checked
      ></Checkbox>
    );
    const labelText = container.prop("labelText");
    const helperText = container.prop("helperText");
    expect(container.find("input").prop("aria-label")).to.equal(
      `${labelText} - ${helperText}`
    );
  });

  it("Renders with label", () => {
    container = Enzyme.mount(
      <Checkbox
        //add other props here
        id="test_id"
        labelText="Hello"
        showLabel={true}
      ></Checkbox>
    );
    expect(container.find("label").exists()).to.equal(true);
    expect(container.find("input").props()).not.to.have.property("aria-label");
    const checkboxId = container.prop("id");
    expect(container.find("label").prop("htmlFor")).to.equal(checkboxId);
  });

  it("Renders with visible helper text", () => {
    container = Enzyme.mount(
      <Checkbox
        //add other props here
        id="test_id"
        labelText="Hello"
        showLabel={true}
        helperText="The helper text."
        errorText="The error text."
        errored={false}
      ></Checkbox>
    );
    expect(container.find(".helper-text").text()).to.equal("The helper text.");
  });

  it("Renders with visible error text", () => {
    container = Enzyme.mount(
      <Checkbox
        //add other props here
        id="test_id"
        labelText="Hello"
        showLabel={true}
        helperText="The helper text."
        errorText="The error text."
        errored={true}
      ></Checkbox>
    );
    expect(container.find(".helper-text").text()).to.equal("The error text.");
  });

  it("Calls a UUID generation method if no ID is passed as a prop", () => {
    container = Enzyme.mount(
      <Checkbox labelText="Hello" showLabel={true}></Checkbox>
    );
    expect(container.find("input").props()).to.have.property("id");

    expect(generateUUIDSpy.called);
  });

  it("The 'checked' attribute can set properly", () => {
    const onChange = stub();
    const container = Enzyme.mount(
      <Checkbox
        id="inputID-attributes"
        labelText="Hello"
        showLabel={false}
        attributes={{
          checked: true,
          "aria-checked": true,
          onChange,
        }}
      ></Checkbox>
    );

    const input = container.find("input");

    expect(input.prop("checked")).to.equal(true);
    expect(input.prop("aria-checked")).to.equal(true);
  });

  it("Passes the ref to the input element", () => {
    const ref = React.createRef<HTMLInputElement>();
    const container = Enzyme.mount(
      <Checkbox
        id="inputID-attributes"
        ref={ref}
        labelText="Hello"
        showLabel={false}
      ></Checkbox>
    );
    expect(container.find("input").instance()).to.equal(ref.current);
  });

  it("Renders HTML attributes passed through the `attributes` prop", () => {
    const onChangeSpy = stub();
    const onBlurSpy = stub();
    container = Enzyme.mount(
      <Checkbox
        id="inputID-attributes"
        attributes={{
          onChange: onChangeSpy,
          onBlur: onBlurSpy,
          tabIndex: 0,
        }}
        labelText="Hello"
        showLabel={true}
      ></Checkbox>
    );
    expect(container.find("input").prop("tabIndex")).to.equal(0);
    expect(onChangeSpy.callCount).to.equal(0);
    container.find("input").simulate("change");
    expect(onChangeSpy.callCount).to.equal(1);
    expect(onBlurSpy.callCount).to.equal(0);
    container.find("input").simulate("blur");
    expect(onBlurSpy.callCount).to.equal(1);
  });
});
