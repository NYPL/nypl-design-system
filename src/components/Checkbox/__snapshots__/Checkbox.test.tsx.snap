// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Renders the UI snapshot correctly 1`] = `
<label
  className="chakra-checkbox css-1cgy6rg"
  onClick={[Function]}
>
  <input
    aria-disabled={false}
    aria-invalid={false}
    checked={false}
    className="chakra-checkbox__input"
    disabled={false}
    id="inputID"
    name="default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    required={false}
    style={
      Object {
        "border": "0px",
        "clip": "rect(0px, 0px, 0px, 0px)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0px",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    <svg
      className="chakra-icon css-1vfv7ic"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <path
        d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="chakra-checkbox__label css-1oeb2oe"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    Test Label
  </span>
</label>
`;

exports[`Checkbox Renders the UI snapshot correctly 2`] = `
<label
  className="chakra-checkbox css-1cgy6rg"
  data-checked=""
  onClick={[Function]}
>
  <input
    aria-disabled={false}
    aria-invalid={false}
    checked={true}
    className="chakra-checkbox__input"
    disabled={false}
    id="checkbox-checked"
    name="default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    required={false}
    style={
      Object {
        "border": "0px",
        "clip": "rect(0px, 0px, 0px, 0px)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0px",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    data-checked=""
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    <svg
      className="chakra-icon css-oi1lnb"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <path
        d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="chakra-checkbox__label css-1oeb2oe"
    data-checked=""
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    Test Label
  </span>
</label>
`;

exports[`Checkbox Renders the UI snapshot correctly 3`] = `
<label
  className="chakra-checkbox css-1cgy6rg"
  data-checked=""
  onClick={[Function]}
>
  <input
    aria-disabled={false}
    aria-invalid={false}
    checked={true}
    className="chakra-checkbox__input"
    disabled={false}
    id="checkbox-checked"
    name="default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    required={false}
    style={
      Object {
        "border": "0px",
        "clip": "rect(0px, 0px, 0px, 0px)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0px",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    data-checked=""
    data-indeterminate=""
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "height": "100%",
          "justifyContent": "center",
          "transform": "none",
        }
      }
    >
      <svg
        className="css-o8vbdw"
        style={
          Object {
            "opacity": 1,
            "stroke": "currentColor",
            "strokeWidth": 4,
            "transform": "none",
          }
        }
        viewBox="0 0 24 24"
      >
        <line
          x1="21"
          x2="3"
          y1="12"
          y2="12"
        />
      </svg>
    </div>
  </span>
  <span
    className="chakra-checkbox__label css-1oeb2oe"
    data-checked=""
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    Test Label
  </span>
</label>
`;

exports[`Checkbox Renders the UI snapshot correctly 4`] = `
<label
  className="chakra-checkbox css-1cgy6rg"
  onClick={[Function]}
>
  <input
    aria-disabled={false}
    aria-invalid={false}
    checked={false}
    className="chakra-checkbox__input"
    disabled={false}
    id="checkbox-required"
    name="default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    required={true}
    style={
      Object {
        "border": "0px",
        "clip": "rect(0px, 0px, 0px, 0px)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0px",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    <svg
      className="chakra-icon css-1vfv7ic"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <path
        d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="chakra-checkbox__label css-1oeb2oe"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    Test Label
  </span>
</label>
`;

exports[`Checkbox Renders the UI snapshot correctly 5`] = `
<label
  className="chakra-checkbox css-1cgy6rg"
  data-invalid=""
  onClick={[Function]}
>
  <input
    aria-disabled={false}
    aria-invalid={true}
    checked={false}
    className="chakra-checkbox__input"
    disabled={false}
    id="checkbox-invalid"
    name="default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    required={false}
    style={
      Object {
        "border": "0px",
        "clip": "rect(0px, 0px, 0px, 0px)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0px",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    data-invalid=""
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    <svg
      className="chakra-icon css-1vfv7ic"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <path
        d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="chakra-checkbox__label css-1oeb2oe"
    data-invalid=""
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    Test Label
  </span>
</label>
`;

exports[`Checkbox Renders the UI snapshot correctly 6`] = `
<label
  className="chakra-checkbox css-1cgy6rg"
  data-disabled=""
  onClick={[Function]}
>
  <input
    aria-disabled={true}
    aria-invalid={false}
    checked={false}
    className="chakra-checkbox__input"
    disabled={true}
    id="checkbox-disabled"
    name="default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    required={false}
    style={
      Object {
        "border": "0px",
        "clip": "rect(0px, 0px, 0px, 0px)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0px",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    data-disabled=""
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    <svg
      className="chakra-icon css-1vfv7ic"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <path
        d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="chakra-checkbox__label css-1oeb2oe"
    data-disabled=""
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    Test Label
  </span>
</label>
`;

exports[`Checkbox Renders the UI snapshot correctly 7`] = `
<label
  className="chakra-checkbox css-vrx9r6"
  onClick={[Function]}
>
  <input
    aria-disabled={false}
    aria-invalid={false}
    checked={false}
    className="chakra-checkbox__input"
    disabled={false}
    id="checkbox-chakra"
    name="default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    required={false}
    style={
      Object {
        "border": "0px",
        "clip": "rect(0px, 0px, 0px, 0px)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0px",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    <svg
      className="chakra-icon css-1vfv7ic"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <path
        d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="chakra-checkbox__label css-1oeb2oe"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    Test Label
  </span>
</label>
`;

exports[`Checkbox Renders the UI snapshot correctly 8`] = `
<label
  className="chakra-checkbox css-1cgy6rg"
  data-testid="testid"
  onClick={[Function]}
>
  <input
    aria-disabled={false}
    aria-invalid={false}
    checked={false}
    className="chakra-checkbox__input"
    disabled={false}
    id="checkbox-props"
    name="default"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    required={false}
    style={
      Object {
        "border": "0px",
        "clip": "rect(0px, 0px, 0px, 0px)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0px",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="chakra-checkbox__control css-dnty2r"
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
  >
    <svg
      className="chakra-icon css-1vfv7ic"
      focusable={false}
      viewBox="0 0 24 24"
    >
      <path
        d="M8.795 15.875l-4.17-4.17-1.42 1.41 5.59 5.59 12-12-1.41-1.41-10.59 10.58z"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    className="chakra-checkbox__label css-1oeb2oe"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    Test Label
  </span>
</label>
`;
